### PR TITLE
8314629: Generational ZGC: Clearing All SoftReferences log line lacks GCId

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

This patch adds the GCId to the log of soft reference policy. I move the related code to the start of the old generational GC which needs  to add a parameter `ZDriverRequest` in the method `ZDriverMajor::collect_old`. An alternative, which doesn't need to add a parameter, is only moving the related code to the start of the major GC, concretely the method `ZDriverMajor::gc`. But considering the references only be cleared in old generational GC, I make such decision. Waiting for your opinion on it.

Thanks for taking the time to review.

Best Regards,
-- Guoxiong

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314629](https://bugs.openjdk.org/browse/JDK-8314629): Generational ZGC: Clearing All SoftReferences log line lacks GCId (**Bug** - P4)

### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17196/head:pull/17196` \
`$ git checkout pull/17196`

Update a local copy of the PR: \
`$ git checkout pull/17196` \
`$ git pull https://git.openjdk.org/jdk.git pull/17196/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17196`

View PR using the GUI difftool: \
`$ git pr show -t 17196`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17196.diff">https://git.openjdk.org/jdk/pull/17196.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17196#issuecomment-1870852437)

### CODE REVIEW

**Review by fisk — CHANGES_REQUESTED** [review]
This makes me a bit nervous. I think this introduces behavioural changes and not just logging changes. By moving the decision making of whether to clear soft references or not past the mark start pause that bumps the sequence number of the two generations, I think you are subtly introducing a race condition that can yield premature OOM issues. If you want to change the logging here, I think we need to find a different way. Perhaps decouple the decision making and logging so we can log wherever we want to without changing the behaviour.

**Comment by lgxbslgx** [comment]
> This makes me a bit nervous. I think this introduces behavioural changes and not just logging changes. By moving the decision making of whether to clear soft references or not past the mark start pause that bumps the sequence number of the two generations, I think you are subtly introducing a race condition that can yield premature OOM issues. If you want to change the logging here, I think we need to find a different way. Perhaps decouple the decision making and logging so we can log wherever we want to without changing the behaviour.

@fisk I missed that the young GC also uses such info to discover references (but only the old generation processes references ). Thanks for reminding me.

What about moving the related code to the start of the method `ZDriverMajor::gc`? The alternative I mentioned.
It means that the tag `O`, shown below, will not be outputted.
```
[4781.774s][info][gc,ref] GC(1239) O: Clearing All SoftReferences
```

**Comment by fisk** [comment]
> > This makes me a bit nervous. I think this introduces behavioural changes and not just logging changes. By moving the decision making of whether to clear soft references or not past the mark start pause that bumps the sequence number of the two generations, I think you are subtly introducing a race condition that can yield premature OOM issues. If you want to change the logging here, I think we need to find a different way. Perhaps decouple the decision making and logging so we can log wherever we want to without changing the behaviour.
> 
> 
> 
> @fisk I missed that the young GC also uses such info to discover references (but only the old generation processes references ). Thanks for reminding me.
> 
> 
> 
> What about moving the related code to the start of the method `ZDriverMajor::gc`? The alternative I mentioned.
> 
> It means that the tag `O`, shown below, will not be outputted.
> 
> ```
> 
> [4781.774s][info][gc,ref] GC(1239) O: Clearing All SoftReferences
> 
> ```
> 
> 

I think I would prefer if the function that sets the soft ref policy just didn't log anything, and we put the logging statement separately where we want it instead. That way I think we can get the log we want without having to change the behaviour. What do you think?

**Comment by lgxbslgx** [comment]
> I think I would prefer if the function that sets the soft ref policy just didn't log anything, and we put the logging statement separately where we want it instead. That way I think we can get the log we want without having to change the behaviour. What do you think?

If we don't log in the method `ZReferenceProcessor::set_soft_reference_policy` and delay the log to a later stage, such as refenrence processing stage, it is hard to get the information about whether we should clear all the soft references. What we can do is judge whether the field `ZReferenceProcessor::_soft_reference_policy` is `AlwaysClearPolicy` or not. Such judgement looks much like RTTI [1] which is disabled in hotspot. In order to obey the dev rule, we can add a virtual method, may named `kind`, to the class `ReferencePolicy` and its subclasses which can avoid RTTI and can judge their types (just like the method `CollectedHeap::kind`). Is my understanding right? If so, it seems not worth to add such method only in order to output one line log.

[1] https://github.com/openjdk/jdk/blob/master/doc/hotspot-style.md#rtti-runtime-type-information

**Comment by fisk** [comment]
I guess you could add a bool member variable to the reference processor to remember the choice.

**Comment by lgxbslgx** [comment]
> I guess you could add a bool member variable to the reference processor to remember the choice.

Thanks for your guide. I updated the code just now.

File: src/hotspot/share/gc/z/zReferenceProcessor.cpp
Comment (fisk) [inline src/hotspot/share/gc/z/zReferenceProcessor.cpp:129]: I would prefer if this assignment was moved above the if and that you assign the field directly the value of the clear argument, instead of having two assignments.

File: src/hotspot/share/gc/z/zReferenceProcessor.hpp
Comment (fisk) [inline src/hotspot/share/gc/z/zReferenceProcessor.hpp:46]: This name doesn't capture that it has something to do with soft references. I think _clear_all_soft_refs is a better name.

File: src/hotspot/share/gc/z/zReferenceProcessor.hpp
Comment (fisk) [inline src/hotspot/share/gc/z/zReferenceProcessor.hpp:45]: This comment confuses me more than it helps. I don't care what the type of the current soft reference policy is, as much as I care about its behaviour. Using RTTI to determine this behaviour based on the type of the current soft ref policy is just one alternative implementation for inspecting the behaviour. I don't think it needs to be mentioned.

**Review by fisk — CHANGES_REQUESTED** [review]
Maybe some small changes...

File: src/hotspot/share/gc/z/zReferenceProcessor.cpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/z/zReferenceProcessor.cpp]: Fixed.

File: src/hotspot/share/gc/z/zReferenceProcessor.hpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/z/zReferenceProcessor.hpp]: Renamed.

File: src/hotspot/share/gc/z/zReferenceProcessor.hpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/z/zReferenceProcessor.hpp]: Fixed. I remove the comment.

**Review by fisk — APPROVED** [review]
Looks good.

**Comment by lgxbslgx** [comment]
Waiting for another review.

**Review by tschatzl — APPROVED** [review]

**Comment by lgxbslgx** [comment]
@fisk @tschatzl Thanks for your reviews. Integrating.

/integrate

**Pull request closed without merging**
